### ISSUE TITLE
2022.2.3 Falsely Indicates Unreachable Code

### ISSUE THREAD

**Issue by Ronocback-Focal** [microsoft/pylance-release#2402]
-   Language Server version: 2022.2.3
-   OS and version: Windows 10 Pro 10.0.19041 Build 19041
-   Python version (& distribution if applicable, e.g. Anaconda): 3.8.3 Anaconda

## Expected behaviour

raygun4py middleware attaching not causing code that follows to display as unreachable.

## Actual behaviour

Raygun provider attachment falsely causes following lines to be marked as unreachable as depicted in screenshot.

![image](https://user-images.githubusercontent.com/99968354/154682404-b22d8e7c-9804-49f9-8da6-71f0d39560bc.png)

Have confirmed that lines following the raygun attachment run as expected despite pylance linting. 

Rolling back pylance to version 2022.2.1 fixes this issue.

Code Snippet:
```python
import flask, os
from flask_cors import CORS
from raygun4py.middleware import flask as rg

#import routes for API
from display import display_blueprint
from qa_backend import qa_backend_blueprint
from login_backend import user_login_blueprint

app = flask.Flask(__name__)
rg.Provider(app, os.environ.get("RAYGUN_KEY")).attach()
app.register_blueprint(display_blueprint)
app.register_blueprint(qa_backend_blueprint)
app.register_blueprint(user_login_blueprint)
CORS(app)

@app.route('/backend/health')
def ping():
    #ECS endpoint, checking condition of container
    return flask.Response(response='Alive!', status=200)

```

**Comment by erictraut** [comment]
I'm not able to repro the problem. It's probably because your code snippet is referencing a bunch of other modules, including (apparently) some internal modules that I don't have access to. Could you try to create a more minimal, self-contained example? If the repro requires the use of third-party packages (e.g. raygun4py), then I'll need to know which versions of those packages you have installed.

**Comment by unit254** [comment]
Language Server version: 2022.2.3
OS and version: Ubuntu 20.04.4 LTS
Python version: 3.8.10 
pyautogui 0.9.53
cv2 4.5.5

I have the same problem but with pyautogui.

![image](https://user-images.githubusercontent.com/5782616/154805140-26fed69b-a804-4b92-bfa4-245b5deef3da.png)

After going back to 2022.2.1
![image](https://user-images.githubusercontent.com/5782616/154805528-04c92f8f-e347-4458-abc5-0e72d6d44626.png)

**Comment by erictraut** [comment]
@unit254, thanks for the additional report. (For future reference, it's useful if you can paste the code into the post so we don't need to retype it from a screen shot.)

Unfortunately, I'm not able to repro that either. I installed pyautogui version 0.9.53 as indicated, and here's what I see:

<img width="346" alt="Screen Shot 2022-02-19 at 7 52 18 AM" src="https://user-images.githubusercontent.com/7040122/154806010-d0e14370-0529-4251-8baf-a93eec12deef.png">

When you hover over `screenshot` in the statement `pyautogui.screenshot(...)`, what type is displayed? Here's what I see:

<img width="418" alt="Screen Shot 2022-02-19 at 7 55 48 AM" src="https://user-images.githubusercontent.com/7040122/154806234-24b34409-8045-407b-a929-f6df0cd7bd44.png">

I suspect that you are seeing something different. If so, please right click on `screenshot` and choose `Go To Declaration`. That should take you to the place where the `screenshot` method is declared. Which file contains this declaration?

**Comment by mebeim** [comment]
@erictraut The issue seems hard to reproduce. I also cannot reproduce it with @unit254's code, but I have another example that has this same problem, using [`pwntools`](https://github.com/elcorto/pwtools) (`pip install pwntools`). Sorry if this is yet another library to install just to test the bug, but I am unable to find a decent minimal example with just standard library modules.

Here's the code that reproduces the issue for me:

```python3
#!/usr/bin/env python3

import pwn

r = pwn.connect('google.com', '80')

def func():
	global r
	r.sendline(b'GET / HTTP/1.1\n\r')

	print('Pylace says this code is unreachable!')
	return 123

func()
print('Pylace says this code is unreachable!')
```

My environment info:

-   Language Server version: 2022.2.3
-   OS and version: linux x64
-   Python version: 3.7.3 
-   python.analysis.indexing: null
-   python.analysis.typeCheckingMode: off
-   Pwntools version: 4.7.0

**Comment by erictraut** [comment]
@mebeim, the issue with pwntools has already been reported in [this issue](https://github.com/microsoft/pylance-release/issues/1847). Fixing this will require either a type stub file or a change in the pwntools library.

**Comment by mebeim** [comment]
@erictraut oh I see, thank you for the link!

**Comment by erictraut** [comment]
@unit254, I just figured out why I wasn't able to repro the problem. I had disabled the `useLibraryCodeForTypes` setting to repro another bug, and I forgot to enable it again. So that was my bad.

The problem in this case is that `pyautogui` is importing the symbol `screenshot` from another library called `pyscreeze`, but it's doing so using a bit of dynamic code that wraps the import in a try/except block. If the import fails, it replaces the `screenshot` with an implementation that raises an exception. Pylance uses static analysis, so it struggles with dynamic behaviors like this. After all, it can't predict statically when the import will succeed. The heuristic it uses when resolving symbols in a module is to always use the last declaration. Normally that heuristic produces the best results, but in this case it's choosing the declaration in the "except" clause rather than the import statement in the "try" block.

I can think of several workarounds:
1. If the maintainer of pyautogui is willing to make a small change to make the package work better with static type checkers, I can suggest a small change that will fix the problem.
2. This library doesn't contain any type information; perhaps there's a type stub package available for it
3. We could look at adjusting the heuristics within pylance so it picks the other declaration in this case; I worry that this change could break more cases than it fixes though, so we'd need to be very cautious

**Comment by mebeim** [comment]
@erictraut is there any plan to add a configuration option for Pylance to avoid doing unreachable code analysis or even just avoid reporting unreachable code? Couldn't find any such option listed in the README. It seems like this would be a better idea than trying to fit the heuristic to every library (most likely impossible) or get specific library maintainers to patch their code (which could be tricky or sometimes even impossible).

**Comment by erictraut** [comment]
Disabling this analysis would result in incorrect type analysis in many cases, so you'd be trading a small number of false positives for a much larger number of false positives if you disabled this. So I don't think that's the answer.

**Comment by mebeim** [comment]
@erictraut I see. Didn't know those two things were dependent on each other, makes sense.

**Comment by Ronocback-Focal** [comment]
> I'm not able to repro the problem. It's probably because your code snippet is referencing a bunch of other modules, including (apparently) some internal modules that I don't have access to. Could you try to create a more minimal, self-contained example? If the repro requires the use of third-party packages (e.g. raygun4py), then I'll need to know which versions of those packages you have installed.

Thanks for getting back to me.

My raygun4py version is 4.3.1 and my flask version is 1.1.2. 

Self-container code example:

```python
import flask, os
from raygun4py.middleware import flask as rg


app = flask.Flask(__name__)
rg.Provider(app, os.environ.get("RAYGUN_KEY")).attach()

@app.route('/backend/health')
def ping():
    #ECS endpoint, checking condition of contianer
    return flask.Response(response='Alive!', status=200)

```

**Comment by erictraut** [comment]
@Ronocback-Focal, thanks for the updated code sample. I'm able to repro the problem. The underlying cause is similar to the issue reported by @unit254. In this case, the `flask.signals` library is attempting to import from a library called `blinker`. The import is performed within a `try` statement. If the import fails, the `except` logic is providing an implementation that raises an exception if called.

I've created a [tracking bug](https://github.com/microsoft/pyright/issues/3093) in the pyright project.

**Comment by erictraut** [comment]
I've improved the heuristics in pyright. It addresses the issue reported with `raygun4py ` and `pyautogui`. This change will be included in the next release of pylance.

**Comment by erictraut** [comment]
If someone sees another instance of "unreachable code" and happens to come across this bug report, please open a new issue with details rather than reusing this same issue. Each of these cases is likely to involve a different set of circumstances even if the symptoms are similar, and it's useful if they are tracked separately.

**Comment by debonte** [comment]
This issue has been fixed in version 2022.2.4, which we've just released. You can find the changelog here: [CHANGELOG.md](https://github.com/microsoft/pylance-release/blob/main/CHANGELOG.md#202224-23-February-2022)
